graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphqlService.java

```diff
@@ -30,49 +31,86 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 
+import com.linecorp.armeria.common.HttpRequest;
 import com.linecorp.armeria.common.HttpResponse;
 import com.linecorp.armeria.common.HttpStatus;
 import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.annotation.Nullable;
 import com.linecorp.armeria.common.graphql.protocol.GraphqlRequest;
+import com.linecorp.armeria.common.websocket.WebSocket;
+import com.linecorp.armeria.common.websocket.WebSocketWriter;
 import com.linecorp.armeria.internal.server.graphql.protocol.GraphqlUtil;
 import com.linecorp.armeria.server.ServiceRequestContext;
 import com.linecorp.armeria.server.graphql.protocol.AbstractGraphqlService;
+import com.linecorp.armeria.server.websocket.IWebSocketService;
+import com.linecorp.armeria.server.websocket.WebSocketService;
+import com.linecorp.armeria.server.websocket.WebSocketServiceHandler;
 
 import graphql.ExecutionInput;
 import graphql.ExecutionResult;
 import graphql.GraphQL;
+import graphql.execution.ExecutionId;
 
-final class DefaultGraphqlService extends AbstractGraphqlService implements GraphqlService {
-
+final class DefaultGraphqlService extends AbstractGraphqlService
+    implements GraphqlService, IWebSocketService, WebSocketServiceHandler, GraphqlExecutor {
     private static final Logger logger = LoggerFactory.getLogger(DefaultGraphqlService.class);
 
     private final GraphQL graphQL;
 
     private final Function<? super ServiceRequestContext,
                            ? extends DataLoaderRegistry> dataLoaderRegistryFunction;
 
+    /**
+     * Used for web sockets where there is no request context.
+     */
+    @Nullable
+    private Supplier<? extends DataLoaderRegistry> dataLoaderRegistrySupplier;
```

Comment:
A `GraphQlService` is shared by multiple connections/requests/event loops so I'm concerned that this field will suffer from race conditions (a registry for requestA may be used for requestB)

What do you think of adding the `DefaultClientRequestContext`  to the `executeGraphql` method parameter and using the provided `dataLoaderRegistryFunction` instead?

e.g.
```java
 interface GraphqlExecutor {
-    ExecutionResult executeGraphql(ExecutionInput.Builder builder);
+    ExecutionResult executeGraphql(Builder builder, ServiceRequestContext ctx);
```

```java
+    private final ServiceRequestContext ctx;
 
     GraphqlWSSubProtocol(ServiceRequestContext ctx, GraphqlExecutor executor) {
         upgradeCtx = GraphqlServiceContexts.graphqlContext(ctx);
         graphqlExecutor = executor;
+        this.ctx = ctx;
     }
```

```java
final ExecutionResult executionResult = graphqlExecutor.executeGraphql(executionInput, ctx);
```

Reply:
Functionally equivalent to what I did with the provider I think but if you think it's clearer then sure I can change it.

Note that the supplier was defined as such:

```
protected HttpResponse doWebSocketUpgrade(ServiceRequestContext ctx, HttpRequest req) throws Exception {
        if (webSocketService != null) {
            dataLoaderRegistrySupplier = () -> dataLoaderRegistryFunction.apply(ctx);
```

Which used the `dataLoaderRegistryFunction` with the same `ctx` as was used to establish the socket, just like your version does.

Reply:
Rather than clarity, I was worried about race conditions since `doWebSocketUpgrade` can be called concurrently by different requests.

i.e.
- RequestA sets `dataLoaderRegistrySupplier` with ctxA
- RequestB sets `dataLoaderRegistrySupplier` with ctxB
- Later, RequestA uses ctxB